### PR TITLE
done SP practise 1part

### PR DESCRIPTION
**Pull request by khorkovenko**
Provide in comments how to use liquidbase.

### CODE REVIEW

File: store/src/main/java/com/book/store/config/SecurityConfig.java [hunk 1]
```diff
@@ -31,8 +29,13 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
...
                         authorizeRequests
-                                .requestMatchers("/auth/**", "/swagger-ui/**")
-                                .permitAll()
+                                .requestMatchers(
+                                        "/auth/**",
+                                        "/swagger-ui/**",
+                                        "/api/categories",
+                                        "/api/categories/**/books"
+                                )
+                                .hasRole("ADMIN")
```
Comment (Rommelua) [inline store/src/main/java/com/book/store/config/SecurityConfig.java]: revert this change

File: store/src/main/java/com/book/store/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+@RequestMapping("/api/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @Operation(summary = "Create category")
+    public ResponseEntity<CategoryDto> createCategory(@RequestBody CategoryDto categoryDto) {
```
Comment (Rommelua) [inline store/src/main/java/com/book/store/controller/CategoryController.java]:
```suggestion
    @Operation(summary = "Create category")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<CategoryDto> createCategory(@RequestBody CategoryDto categoryDto) {
```

File: store/src/main/java/com/book/store/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotBlank
+    private String name;
+
+    @Column(name = "deleted")
+    private boolean deleted;
```
Comment (Rommelua) [inline store/src/main/java/com/book/store/model/Category.java]:
```suggestion
    @Column(name = "is_deleted")
    private boolean isDeleted;
```

File: store/src/main/java/com/book/store/model/Category.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package com.book.store.model;
+
+import jakarta.persistence.*;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Entity
+@Data
+public class Category {
```
Comment (Rommelua) [inline store/src/main/java/com/book/store/model/Category.java]: @SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?;")
@Where(clause = "is_deleted = false")
image

File: store/src/main/java/com/book/store/service/impl/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return categoryMapper.toDto(updatedCategory);
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        Optional<Category> categoryOptional = categoryRepository.findById(id);
+        categoryOptional.ifPresent(category -> {
+            category.setDeleted(true);
+            categoryRepository.save(category);
+        });
```
Comment (Rommelua) [inline store/src/main/java/com/book/store/service/impl/CategoryServiceImpl.java]:
```suggestion
        categoryRepository.deleteById(id)
```

File: store/src/main/java/com/book/store/config/SecurityConfig.java [hunk 6]
```diff
@@ -31,8 +29,11 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .addFilterBefore(new JwtAuthenticationFilter(jwtUtil, userDetailsService), UsernamePasswordAuthenticationFilter.class)
                 .authorizeHttpRequests(authorizeRequests ->
                         authorizeRequests
-                                .requestMatchers("/auth/**", "/swagger-ui/**")
-                                .permitAll()
+                                .requestMatchers(
+                                        "/auth/**",
+                                        "/swagger-ui/**"
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/config/SecurityConfig.java:38]: permit `/error`. 

User should have access to error page in case of exceptions.

Comment (sokimaaa) [inline store/src/main/java/com/book/store/config/SecurityConfig.java:38]: why /auth permitted only for admins?

File: store/src/main/java/com/book/store/controller/BookController.java [hunk 7]
```diff
@@ -77,14 +77,15 @@ public ResponseEntity<Void> deleteBook(@PathVariable Long id) {
     }
 
     @GetMapping("/search")
-    @Operation(summary = "Search books based on criteria")
-    public ResponseEntity<List<BookDto>> searchBooks(
+    @Operation(summary = "Search books based on criteria with pagination and sorting")
+    public ResponseEntity<Page<BookDto>> searchBooks(
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/controller/BookController.java]: Page is a database object that is used for extracting partial data. 
Collection is used to transfer data. 

Replace return type with Collection, e.g. list

```suggestion
    public ResponseEntity<List<BookDto>> searchBooks(
```

File: store/src/main/java/com/book/store/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    @Autowired
+    public CategoryServiceImpl(CategoryRepository categoryRepository, CategoryMapper categoryMapper) {
+        this.categoryRepository = categoryRepository;
+        this.categoryMapper = categoryMapper;
+    }
+
+    @Override
+    public Page<CategoryDto> findAll(Pageable pageable) {
+        Page<Category> categoryPage = categoryRepository.findAll(pageable);
+        return categoryPage.map(categoryMapper::toDto);
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/service/impl/CategoryServiceImpl.java:22]: inline
```suggestion
        return categoryRepository.findAll(pageable).map(categoryMapper::toDto);
```

File: store/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 9]
```diff
@@ -0,0 +1,2 @@
+databaseChangeLog:
+  -
```
Comment (sokimaaa) [inline store/src/main/resources/db/changelog/db.changelog-master.yaml]: why empty? fill with liquibase scripts

File: store/src/main/java/com/book/store/config/SecurityConfig.java (hunk 6 above)
Comment (Rommelua) [reply store/src/main/java/com/book/store/config/SecurityConfig.java:38]: not fixed

File: store/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 10]
```diff
@@ -0,0 +1,108 @@
+databaseChangeLog:
+  - changeSet:
+      id: createBookTable
+      author: your_author_name
```
Comment (Rommelua) [inline store/src/main/resources/db/changelog/db.changelog-master.yaml]: see how it should looks like
image
